File: log4j/src/main/java/com/tuuzed/androidx/logging/log4j/Log4jLogcatAppender.java
package com.tuuzed.androidx.logging.log4j;

import org.apache.log4j.AppenderSkeleton;
import org.apache.log4j.Layout;
import org.apache.log4j.Level;
import org.apache.log4j.PatternLayout;
import org.apache.log4j.spi.LoggingEvent;

public class Log4jLogcatAppender extends AppenderSkeleton {

    private final Layout tagLayout = new PatternLayout("%c");
    private final Layout messageLayout = new PatternLayout("%m%n");

    public Log4jLogcatAppender() {
        setLayout(messageLayout);
    }

    @Override
    protected void append(LoggingEvent event) {
        switch (event.getLevel().toInt()) {
            case Level.TRACE_INT:
                if (event.getThrowableInformation() == null) {
                    android.util.Log.v(tagLayout.format(event), getLayout().format(event));
                } else {
                    android.util.Log.v(tagLayout.format(event), getLayout().format(event), event.getThrowableInformation().getThrowable());
                }
                break;
            case Level.DEBUG_INT:
                if (event.getThrowableInformation() == null) {
                    android.util.Log.d(tagLayout.format(event), getLayout().format(event));
                } else {
                    android.util.Log.d(tagLayout.format(event), getLayout().format(event), event.getThrowableInformation().getThrowable());
                }
                break;
            case Level.INFO_INT:
                if (event.getThrowableInformation() == null) {
                    android.util.Log.i(tagLayout.format(event), getLayout().format(event));
                } else {
                    android.util.Log.i(tagLayout.format(event), getLayout().format(event), event.getThrowableInformation().getThrowable());
                }
                break;
            case Level.WARN_INT:
                if (event.getThrowableInformation() == null) {
                    android.util.Log.w(tagLayout.format(event), getLayout().format(event));
                } else {
                    android.util.Log.w(tagLayout.format(event), getLayout().format(event), event.getThrowableInformation().getThrowable());
                }
                break;
            case Level.ERROR_INT:
                if (event.getThrowableInformation() == null) {
                    android.util.Log.e(tagLayout.format(event), getLayout().format(event));
                } else {
                    android.util.Log.e(tagLayout.format(event), getLayout().format(event), event.getThrowableInformation().getThrowable());
                }
                break;
            case Level.FATAL_INT:
                if (event.getThrowableInformation() == null) {
                    android.util.Log.wtf(tagLayout.format(event), getLayout().format(event));
                } else {
                    android.util.Log.wtf(tagLayout.format(event), getLayout().format(event), event.getThrowableInformation().getThrowable());
                }
                break;
        }

    }

    @Override
    public void close() {
    }

    @Override
    public boolean requiresLayout() {
        return true;
    }
}

File: log4j/src/main/java/com/tuuzed/androidx/logger/log4j/Log4jLogcatAppender.kt
package com.tuuzed.androidx.logger.log4j

import android.util.Log
import org.apache.log4j.AppenderSkeleton
import org.apache.log4j.Layout
import org.apache.log4j.Level
import org.apache.log4j.PatternLayout
import org.apache.log4j.spi.LoggingEvent

class Log4jLogcatAppender @JvmOverloads constructor(
    messageLayout: Layout = PatternLayout("%m%n"),
    var tagLayout: Layout = PatternLayout("%c")
) : AppenderSkeleton() {

    init {
        this.setLayout(messageLayout)
    }

    override fun append(le: LoggingEvent) {
        when (le.getLevel()) {
            Level.TRACE -> if (le.throwableInformation != null) {
                Log.v(
                    this.tagLayout.format(le),
                    this.getLayout().format(le),
                    le.throwableInformation.throwable
                )
            } else {
                Log.v(this.tagLayout.format(le), this.getLayout().format(le))
            }
            Level.DEBUG -> if (le.throwableInformation != null) {
                Log.d(
                    this.tagLayout.format(le),
                    this.getLayout().format(le),
                    le.throwableInformation.throwable
                )
            } else {
                Log.d(this.tagLayout.format(le), this.getLayout().format(le))
            }
            Level.INFO -> if (le.throwableInformation != null) {
                Log.i(
                    this.tagLayout.format(le),
                    this.getLayout().format(le),
                    le.throwableInformation.throwable
                )
            } else {
                Log.i(this.tagLayout.format(le), this.getLayout().format(le))
            }
            Level.WARN -> if (le.throwableInformation != null) {
                Log.w(
                    this.tagLayout.format(le),
                    this.getLayout().format(le),
                    le.throwableInformation.throwable
                )
            } else {
                Log.w(this.tagLayout.format(le), this.getLayout().format(le))
            }
            Level.ERROR -> if (le.throwableInformation != null) {
                Log.e(
                    this.tagLayout.format(le),
                    this.getLayout().format(le),
                    le.throwableInformation.throwable
                )
            } else {
                Log.e(this.tagLayout.format(le), this.getLayout().format(le))
            }
            Level.FATAL -> if (le.throwableInformation != null) {
                Log.wtf(
                    this.tagLayout.format(le),
                    this.getLayout().format(le),
                    le.throwableInformation.throwable
                )
            } else {
                Log.wtf(this.tagLayout.format(le), this.getLayout().format(le))
            }
        }

    }

    override fun close() {}
    override fun requiresLayout(): Boolean = true

}